sunspec/core/util.py

```diff
@@ -95,6 +96,12 @@ def data_to_double(data):
         return d[0]
 
 def data_to_str(data):
+
+    # Change the data from bytes string to regular string for python 3
+    # compatibility
+    if sys.version_info > (3,):
+        data = str(data, 'latin-1')
```

Comment:
Sorry for the delay. 

> You shouldn't arbitrarily flip between byte sequences and character sequences. 

It wasn't arbitrary, I'm not familiar with modbus and so I was not confident in changing the processing code. The approach I took then was to change the type of the inputs and outputs to the processing code to make use of the existing code. 

I tried looking at how you implemented [changes](https://github.com/altendky/pysunspec/commit/ffbdd14134da551657d5d0f7b30729ce94498b66) to your repository and I think you approached it by using `from builtins import str`. When I tried implementing this I'm still getting errors. Also I don't think python 2.X has `builtins` as part of the standard library.

```
======================================================================
ERROR: test_client_device (core.test.test_client.TestClientDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/mbmap.py", line 254, in from_xml
    data = struct.pack(str(rlen * 2) + 's', text)
struct.error: argument for 's' must be a bytes object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_client.py", line 45, in test_client_device
    pathlist=self.pathlist)
  File "./sunspec/core/client.py", line 143, in __init__
    self.modbus_device = modbus.ModbusClientDeviceMapped(slave_id, name, pathlist, self)
  File "./sunspec/core/modbus/client.py", line 911, in __init__
    self.modbus_map.from_xml(name, pathlist)
  File "./sunspec/core/modbus/mbmap.py", line 292, in from_xml
    raise ModbusMapError('Error loading %s (%s) at offset %d - %s' % (filename, pathlist, offset, str(e)))
sunspec.core.modbus.mbmap.ModbusMapError: Error loading mbmap_test_device_1.xml (['.', '/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/devices']) at offset 0 - argument for 's' must be a bytes object

======================================================================
ERROR: test_sunspec_client_device_1 (core.test.test_client.TestClientDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/mbmap.py", line 254, in from_xml
    data = struct.pack(str(rlen * 2) + 's', text)
struct.error: argument for 's' must be a bytes object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_client.py", line 60, in test_sunspec_client_device_1
    pathlist=self.pathlist)
  File "./sunspec/core/client.py", line 768, in __init__
    self.device = ClientDevice(device_type, slave_id, name, pathlist, baudrate, parity, ipaddr, ipport, timeout, trace)
  File "./sunspec/core/client.py", line 143, in __init__
    self.modbus_device = modbus.ModbusClientDeviceMapped(slave_id, name, pathlist, self)
  File "./sunspec/core/modbus/client.py", line 911, in __init__
    self.modbus_map.from_xml(name, pathlist)
  File "./sunspec/core/modbus/mbmap.py", line 292, in from_xml
    raise ModbusMapError('Error loading %s (%s) at offset %d - %s' % (filename, pathlist, offset, str(e)))
sunspec.core.modbus.mbmap.ModbusMapError: Error loading mbmap_test_device_1.xml (['.', '/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/devices']) at offset 0 - argument for 's' must be a bytes object

======================================================================
ERROR: test_sunspec_client_device_3 (core.test.test_client.TestClientDevice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/mbmap.py", line 254, in from_xml
    data = struct.pack(str(rlen * 2) + 's', text)
struct.error: argument for 's' must be a bytes object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_client.py", line 164, in test_sunspec_client_device_3
    d = client.SunSpecClientDevice(client.MAPPED, slave_id=1, name='mbmap_test_device_3.xml', pathlist=self.pathlist)
  File "./sunspec/core/client.py", line 768, in __init__
    self.device = ClientDevice(device_type, slave_id, name, pathlist, baudrate, parity, ipaddr, ipport, timeout, trace)
  File "./sunspec/core/client.py", line 143, in __init__
    self.modbus_device = modbus.ModbusClientDeviceMapped(slave_id, name, pathlist, self)
  File "./sunspec/core/modbus/client.py", line 911, in __init__
    self.modbus_map.from_xml(name, pathlist)
  File "./sunspec/core/modbus/mbmap.py", line 292, in from_xml
    raise ModbusMapError('Error loading %s (%s) at offset %d - %s' % (filename, pathlist, offset, str(e)))
sunspec.core.modbus.mbmap.ModbusMapError: Error loading mbmap_test_device_3.xml (['.', '/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/devices']) at offset 0 - argument for 's' must be a bytes object

======================================================================
ERROR: test_data (core.test.test_data.TestData)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_data.py", line 47, in test_data
    raise Exception('XML mismatch: %s %s' % (xml, expected_xml))
Exception: XML mismatch: b'<sunSpecData><d lid="00:00:00:00:00:01" man="Man" mod="Mod" sn="SN" t="2013-12-20T17:29:24Z"><m id="1"><p id="P1">23</p><p id="P2">24</p></m></d></sunSpecData>' <sunSpecData><d lid="00:00:00:00:00:01" man="Man" mod="Mod" sn="SN" t="2013-12-20T17:29:24Z"><m id="1"><p id="P1">23</p><p id="P2">24</p></m></d></sunSpecData>

======================================================================
ERROR: test_modbus_client_device_rtu_read (core.test.test_modbus_client.TestModbusClient)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_client.py", line 44, in test_modbus_client_device_rtu_read
    data = d.read(40000, 2)
  File "./sunspec/core/modbus/client.py", line 513, in read
    return self.client.read(self.slave_id, addr, count, op=op, trace_func=self.trace_func, max_count=self.max_count)
  File "./sunspec/core/modbus/client.py", line 315, in read
    data = self._read(slave_id, addr + read_offset, read_count, op=op, trace_func=trace_func)
  File "./sunspec/core/modbus/client.py", line 231, in _read
    req += struct.pack('>H', computeCRC(req))
  File "./sunspec/core/modbus/client.py", line 995, in computeCRC
    idx = __crc16_table[(crc ^ ord(a)) & 0xff];
TypeError: ord() expected string of length 1, but int found

======================================================================
ERROR: test_modbus_client_device_rtu_write (core.test.test_modbus_client.TestModbusClient)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_client.py", line 66, in test_modbus_client_device_rtu_write
    d.write(40000, 'ABCD')
  File "./sunspec/core/modbus/client.py", line 527, in write
    return self.client.write(self.slave_id, addr, data, trace_func=self.trace_func, max_count=self.max_count)
  File "./sunspec/core/modbus/client.py", line 419, in write
    trace_func=trace_func)
  File "./sunspec/core/modbus/client.py", line 337, in _write
    req += data
TypeError: can't concat bytes to str

======================================================================
ERROR: test_modbus_client_device_tcp_read (core.test.test_modbus_client.TestModbusClient)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/client.py", line 683, in _read
    self.socket.sendall(req)
  File "./sunspec/core/test/fake/socket.py", line 63, in sendall
    self.out_buf += data
TypeError: Can't convert 'bytes' object to str implicitly

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_client.py", line 85, in test_modbus_client_device_tcp_read
    data = d.read(40000, 2)
  File "./sunspec/core/modbus/client.py", line 753, in read
    data = self._read(addr + read_offset, read_count, op=op)
  File "./sunspec/core/modbus/client.py", line 685, in _read
    raise ModbusClientError('Socket write error: %s' % str(e))
sunspec.core.modbus.client.ModbusClientError: Socket write error: Can't convert 'bytes' object to str implicitly

======================================================================
ERROR: test_modbus_client_device_tcp_write (core.test.test_modbus_client.TestModbusClient)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_client.py", line 107, in test_modbus_client_device_tcp_write
    d.write(40000, 'ABCD')
  File "./sunspec/core/modbus/client.py", line 850, in write
    self._write(addr + write_offset, data[start:end])
  File "./sunspec/core/modbus/client.py", line 779, in _write
    req += data
TypeError: can't concat bytes to str

======================================================================
ERROR: test_modbus_mbmap_from_xml_element (core.test.test_modbus_mbmap.TestModbusMap)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/mbmap.py", line 254, in from_xml
    data = struct.pack(str(rlen * 2) + 's', text)
struct.error: argument for 's' must be a bytes object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_mbmap.py", line 78, in test_modbus_mbmap_from_xml_element
    m1.from_xml(element=root)
  File "./sunspec/core/modbus/mbmap.py", line 292, in from_xml
    raise ModbusMapError('Error loading %s (%s) at offset %d - %s' % (filename, pathlist, offset, str(e)))
sunspec.core.modbus.mbmap.ModbusMapError: Error loading None (None) at offset 0 - argument for 's' must be a bytes object

======================================================================
ERROR: test_modbus_mbmap_from_xml_file (core.test.test_modbus_mbmap.TestModbusMap)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./sunspec/core/modbus/mbmap.py", line 254, in from_xml
    data = struct.pack(str(rlen * 2) + 's', text)
struct.error: argument for 's' must be a bytes object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/test_modbus_mbmap.py", line 47, in test_modbus_mbmap_from_xml_file
    m1.from_xml('mbmap_test_device_1.xml', self.pathlist)
  File "./sunspec/core/modbus/mbmap.py", line 292, in from_xml
    raise ModbusMapError('Error loading %s (%s) at offset %d - %s' % (filename, pathlist, offset, str(e)))
sunspec.core.modbus.mbmap.ModbusMapError: Error loading mbmap_test_device_1.xml (['.', '/Users/james/git_repos/pysunspec_work/pysunspec/sunspec/core/test/devices']) at offset 0 - argument for 's' must be a bytes object

----------------------------------------------------------------------
Ran 22 tests in 0.234s

FAILED (errors=10)
```